File: 2020/02/password_philosophy.py
# password_philosophy.py
# Callum Howard, 2020

from sys import stdin
import re

def solve(line):
    range, letter, password = line.split()
    min, max = map(int, range.split('-'))
    return min <= sum(l == letter[0] for l in password) <= max

def solve2(line):
    min, max, letter, password = re.match(r'(\d+)-(\d+)\s([\w+]):\s(\w+)', line).groups()
    return int(min) <= sum(l == letter for l in password) <= int(max)

print(sum(solve2(line) for line in stdin.readlines()))

File: 2020/02/password_philosophy.go
// Password Philosophy
// https://adventofcode.com/2020/day/2

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func countLetter(inputLetter rune, input string) int {
	var count = 0
	for _, letter := range input {
		if letter == inputLetter {
			count++
		}
	}
	return count
}

func solve(line string) bool {
	split := strings.Fields(line)
	minMax := strings.Split(split[0], "-")
	min, _ := strconv.Atoi(minMax[0])
	max, _ := strconv.Atoi(minMax[1])
	letter := rune(split[1][0])
	password := split[2]

	count := countLetter(letter, password)
	return min <= count && count <= max
}

func main() {
	var scanner = bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanLines)

	var count = 0
	for scanner.Scan() {
		if solve(scanner.Text()) {
			count++
		}
	}
	fmt.Println(count)
}
